RejectedExecutionException

# RejectedExecutionException

### 1. 背景
今天我们的一个服务偶然抛了以下异常：
```
Exception in thread "main" java.util.concurrent.RejectedExecutionException: Task test.suite.ThreadTest$$Lambda$1/759156157@120d6fe6 rejected from java.util.concurrent.ThreadPoolExecutor@4ba2ca36[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 0]
	at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2047)
	at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:823)
	at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1369)
	at java.util.concurrent.Executors$DelegatedExecutorService.execute(Executors.java:668)
	at test.suite.ThreadTest.test(ThreadTest.java:15)
	at test.suite.ThreadTest.main(ThreadTest.java:9)
```
抛出异常的代码块类似于以下逻辑(按以下逻辑执行，大概率会抛RejectedExecutionException)：

```
public static void main(String[] args){
        for (int i = 0; i < 1000; i++) {
            test();
        }
    }

public static void test(){
  ExecutorService executor = Executors.newSingleThreadExecutor();
  executor.execute(() -> {
    Long now = System.currentTimeMillis();
    Long j = 0L;
    for (int i = 0; i < 10000000; i++) {
    j = j + 1L;
    }
    System.out.println(System.currentTimeMillis() - now);
  });
}
```
首先忽略下面两个坑点：

a.方法内new一个临时线程池，不仅无法复用，浪费资源，而且性能很差(用一个全局的线程池或者直接new一个Thread也比这样好很多)

b.没有主动关闭资源，可能造成资源无法释放。

因为除了以上已知问题，本文主要想把为什么会抛RejectedExecutionException异常搞明白。

### 2.排查过程
最开始怀疑并发调用了test方法，创建了大量的线程池和线程，而线程资源是有限的，在线程资源不足时，新建Executor可能无法初始化线程，因此导致调用失败。

因此本地尝试复现线程创建失败的场景。但是发现线程创建失败，报错如下：
```
Exception in thread "main" java.lang.OutOfMemoryError: unable to create new native thread
	at java.lang.Thread.start0(Native Method)
	at java.lang.Thread.start(Thread.java:714)
	at java.util.concurrent.ThreadPoolExecutor.addWorker(ThreadPoolExecutor.java:950)
	at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1357)
	at test.suite.ThreadTest.test(ThreadTest.java:21)
	at test.suite.ThreadTest.main(ThreadTest.java:12)
```
和RejectedExecutionException的猜测不符。

仔细查看Executors的源码，发现以下内容：

```
 //新建线程池，coreThread和maxThread的个数都是1，无限队列。
    public static ExecutorService newSingleThreadExecutor() {
        return new FinalizableDelegatedExecutorService
            (new ThreadPoolExecutor(1, 1,
                                    0L, TimeUnit.MILLISECONDS,
                                    new LinkedBlockingQueue<Runnable>()));
    
    //该方法对ThreadPoolExecutor进行了简单封装，重写了finalize方法，不用显示调用shutdown，利用gc时可以自动触发。
    static class FinalizableDelegatedExecutorService
        extends DelegatedExecutorService {
        FinalizableDelegatedExecutorService(ExecutorService executor) {
            super(executor);
        }
        protected void finalize() {
            super.shutdown();
        }
```
最初看到以上代码时，发现封装了一层FinalizableDelegatedExecutorService，还不错，就算忘记显式shutdown，资源也能正常释放，也解释了为啥服务没有内存泄露的问题。

但是为什么抛RejectedExecutionException异常，还是没有解决。

会不会调用executor之前，恰好触发了gc，显示调用了finalize，从而调用了shutdown呢？因此进行以下测试：
```
public static void main(String[] args){
        for (int i = 0; i < 1; i++) {
            test();
        }
    }

public static void test(){
  ExecutorService executor = new Executors.newSingleThreadExecutor();
  executor.shutdown();
  executor.execute(() -> {
    Long now = System.currentTimeMillis();
    Long j = 0L;
    for (int i = 0; i < 10000000; i++) {
    j = j + 1L;
    }
    System.out.println(System.currentTimeMillis() - now);
  });
}	
```
发现抛了相同的RejectedExecutionException。

再次进行以下尝试：
```
public static void main(String[] args) {
        for (int i = 0; i < 1000; i++) {
            test();
        }
    }

public static void test(){
  ExecutorService executor = new ThreadPoolExecutor(1, 1,0L, TimeUnit.MILLISECONDS,new LinkedBlockingQueue<Runnable>());
  executor.execute(() -> {
    Long now = System.currentTimeMillis();
    Long j = 0L;
    for (int i = 0; i < 10000000; i++) {
    j = j + 1L;
    }
    System.out.println(System.currentTimeMillis() - now);
  });
  executor.shutdown();
}
```
发现不再抛RejectedExecutionException了，多次尝试都没问题。

### 结论
因此大胆的怀疑以下两行代码被jvm优化了。因为new了executor后，只有紧邻的一行代码使用到了，而jvm在executor.execute()调用时，只要找到了虚方法表的地址，就可以直接释放executor变量；

由于gc比较频繁，恰好gc把executor变量进行了回收，显示调用了finalize()方法，进而调用了shutdown方法，

而此时executor.execute()逻辑才刚开始执行，发现线程池已经被关闭，从而抛RejectedExecutionException异常。
```
ExecutorService executor = new Executors.newSingleThreadExecutor();
executor.execute(() -> {
	...
})
```
#### 最后

以后尽量不要使用Executors中的静态方法，因为有以下两个坑：

1.通过Executors新建的都是无界队列，在极端情况下可能会造成内存溢出，例如调用下游RPC方法被阻塞；

2.FinalizableDelegatedExecutorService中重写的finalize感觉还是有风险的。

推荐优先使用以下方式：


```
ThreadFactory namedThreadFactory = new ThreadFactoryBuilder().setNameFormat("demo-pool-%d").build();

//Common Thread Pool
ExecutorService pool = new ThreadPoolExecutor(5, 200,
0L, TimeUnit.MILLISECONDS,
new LinkedBlockingQueue<Runnable>(1024), namedThreadFactory, new ThreadPoolExecutor.AbortPolicy());

pool.execute(()-> System.out.println(Thread.currentThread().getName()));
pool.shutdown();//gracefully shutdown
```

id: d16e141b728942378315dfc5360ac1a4
parent_id: 4cccde86180d4a45b998f084a85041a8
created_time: 2020-04-15T11:25:43.307Z
updated_time: 2020-04-15T11:11:00.466Z
is_conflict: 0
latitude: 0.00000000
longitude: 0.00000000
altitude: 0.0000
author: 
source_url: 
is_todo: 0
todo_due: 0
todo_completed: 0
source: joplin-desktop
source_application: net.cozic.joplin-desktop
application_data: 
order: 0
user_created_time: 2020-04-15T11:25:43.307Z
user_updated_time: 2020-04-15T11:11:00.466Z
encryption_cipher_text: 
encryption_applied: 0
markup_language: 1
is_shared: 0
type_: 1